src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,688 @@
+            {
+                varDecl = syntaxFacts.GetIdentifierOfVariableDeclarator(expressionDecl).ValueText;
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Gets the method symbol the expression is enclosed within
+        /// </summary>
+        protected IMethodSymbol? GetMethodSymbolFromExpression(SemanticDocument semanticDocument,
+            TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var methodSymbol = (IMethodSymbol)semanticDocument.SemanticModel.GetRequiredEnclosingSymbol(expression.SpanStart, cancellationToken);
+            var syntaxFacts = semanticDocument.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var methodDeclaration = expression.FirstAncestorOrSelf<SyntaxNode>(node => IsContainedInParameterizedDeclaration(node), ascendOutOfTrivia: true);
+            if (methodDeclaration is null)
+            {
+                return null;
+            }
+
+            return (IMethodSymbol)semanticDocument.SemanticModel.GetDeclaredSymbol(methodDeclaration, cancellationToken)!;
+        }
+
+        /// <summary>
+        /// Locates all the call sites of the method that introduced the parameter
+        /// </summary>
+        /// <returns>Dictionary tying all the invocations to each corresponding document</returns>
+        protected static async Task<ImmutableDictionary<Document, List<TInvocationExpressionSyntax>>> FindCallSitesAsync(
+            SemanticDocument document, IMethodSymbol methodSymbol, CancellationToken cancellationToken)
+        {
+            var methodCallSites = new Dictionary<Document, List<TInvocationExpressionSyntax>>();
+            var progress = new StreamingProgressCollector();
+
+            await SymbolFinder.FindReferencesAsync(
+                methodSymbol, document.Document.Project.Solution, progress: progress,
+                documents: null, FindReferencesSearchOptions.Default, cancellationToken).ConfigureAwait(false);
+            var referencedSymbols = progress.GetReferencedSymbols();
+            var referencedLocations = referencedSymbols.SelectMany(referencedSymbol => referencedSymbol.Locations).Distinct().ToImmutableArray()
+                .OrderByDescending(reference => reference.Location.SourceSpan.Start);
+
+            var list = new List<TInvocationExpressionSyntax>();
+            methodCallSites.Add(document.Document, list);
+            foreach (var refLocation in referencedLocations)
+            {
+                if (!methodCallSites.TryGetValue(refLocation.Document, out var invocations))
+                {
+                    invocations = new List<TInvocationExpressionSyntax>();
+                    methodCallSites.Add(refLocation.Document, invocations);
+                }
+
+                invocations.Add((TInvocationExpressionSyntax)refLocation.Location.FindNode(cancellationToken).GetRequiredParent());
```

Comment:
this cast is not safe.  not all locations will be invocations.  for example.  `Func<int, string> f = this.M;`.  PLease test this as this will currently crash.

Reply:
I added a check to see if the SyntaxNode is an InvocationExpression